// @swf/core/src/pltable/test/js/splmTableScrollTestService.js
export const getData = () => {
    let _totalFound = 70;
    let _totalObjectsLoaded = 50;
    let _response = [];
    var _mockObjCnt = 0;
    for( var i = 0; i < _totalObjectsLoaded; i++ ) {
        _response.push( {
            uid: _mockObjCnt,
            props: {
                name: {
                    isNull: false,
                    type: 'STRING',
                    dbValue: 'File ' + _mockObjCnt,
                    uiValue: 'File ' + _mockObjCnt
                },
                date: {
                    isNull: false,
                    type: 'STRING',
                    dbValue: '2018-06-01',
                    uiValue: '2018-06-01',
                    isEditable: true
                },
                dropdownLOV: {
                    isNull: false,
                    type: 'STRING',
                    dbValue: _mockObjCnt,
                    uiValue: _mockObjCnt,
                    isEditable: true,
                    hasLov: true
                },
                aw2_TypedReference: {
                    isNull: false,
                    type: 'OBJECT',
                    dbValue: 'Test',
                    uiValue: 'Test',
                    uiValues: [ 'Test' ],
                    isEditable: true,
                    propertyName: 'aw2_TypedReference',
                    isArray: true,
                    propApi: {

                    }
                }
            }
        } );
        _mockObjCnt++;
    }
    return {
        searchResults: _response,
        totalFound: _totalFound,
        totalObjectsLoaded: _totalObjectsLoaded
    };
};
