// @swf/core/src/pltable/test/js/splmTableMaxRowsToShowService.js
export const getData = () => {
    let _totalFound = 50;
    let _response = [];
    for( var i = 0; i < _totalFound; i++ ) {
        _response.push( {
            uid: i,
            props: {
                name: {
                    isNull: false,
                    type: 'STRING',
                    dbValue: 'File ' + i,
                    uiValue: 'File ' + i
                },
                date: {
                    isNull: false,
                    type: 'STRING',
                    dbValue: '2018-06-01',
                    uiValue: '2018-06-01',
                    isEditable: true
                },
                dropdownLOV: {
                    isNull: false,
                    type: 'STRING',
                    dbValue: i,
                    uiValue: i,
                    isEditable: true,
                    hasLov: true
                },
                aw2_TypedReference: {
                    isNull: false,
                    type: 'OBJECT',
                    dbValue: 'Test',
                    uiValue: 'Test',
                    uiValues: [ 'Test' ],
                    isEditable: true,
                    propertyName: 'aw2_TypedReference',
                    isArray: true,
                    propApi: {

                    }
                }
            }
        } );
    }
    return {
        searchResults: _response,
        totalFound: _totalFound
    };
};
