// @swf/core/src/uielements/src/js/buttonStyles.js
// Copyright (c) 2020 Siemens
/**
 * Define the common styles shared for button and chip
 * @module "js/buttonStyles"
 */

const ACCENT_CAUTION = 'accent-caution';
// Define the map from button type to CSS class
const buttonStyles = {
    base: 'base-button',
    accent: 'base-button', //only applicable for selection chip
    caution: 'caution',
    sole: ACCENT_CAUTION,
    positive: 'positive',
    negative: 'negative',
    chromeless: 'chromeless',
    'accent-high-contrast': 'accent-highContrast',
    'accent-mid-contrast': 'accent-midContrast',
    'accent-positive': 'accent-positive',
    'accent-caution': ACCENT_CAUTION,
    'accent-negative': 'accent-negative',
    'accent-marketing': ACCENT_CAUTION
};

/**
 * Get button style, if button type is undefined or not supported, return the style of default type.
 * @param {String} buttonType the button type
 * @param {String} defaultType the default button type.
 * @returns {String} button CSS class
 */
export let getButtonStyle = function( buttonType ) {
    var btnClass = '';
    if( buttonType && buttonStyles.hasOwnProperty( buttonType ) ) {
        btnClass = buttonStyles[ buttonType ];
    }

    return btnClass;
};

export default { getButtonStyle };
