// @swf/core/src/layout/src/js/awHeaderService.js
// Copyright (c) 2020 Siemens
/**
 * @module js/awHeaderService
 */
import _ from 'lodash';
import AwPromiseService from 'js/awPromiseService';
import AwStateService from 'js/awStateService';
import localeService from 'js/localeService';
import appCtxService from 'js/appCtxService';
import configurationService from 'js/configurationService';
import breadCrumbService from 'js/breadCrumbService';
import logger from 'js/logger';
import { getActiveSublocationTab } from 'js/awSearchSublocationService';

var contextName = 'location.titles';

export let getTitles = function() {
    var output = {};
    var promises = [];

    promises.push( configurationService.getCfg( 'solutionDef' ).then( function( solution ) {
        var browserTitle = solution ? solution.browserTitle : 'Apollo';
        output.browserTitle = browserTitle;
        return output;
    } ) );

    [ 'browserSubTitle', 'headerTitle' ].forEach( function( key ) {
        var property = AwStateService.instance.current.data[ key ];
        if( property ) {
            if( typeof property === 'string' ) {
                output[ key ] = property;
                promises.push( AwPromiseService.instance.when( output ) );
            } else {
                promises.push( localeService.getLocalizedText( property.source, property.key )
                    .then( function( result ) {
                        output[ key ] = result;
                        return output;
                    } ) );
            }
        }
    } );

    return AwPromiseService.instance.all( promises ).then( function() {
        return output;
    } );
};

export const getBreadcrumbConfig = () => {
    return AwStateService.instance.current.data ? AwStateService.instance.current.data.breadcrumbConfig : {};
};

export let updateBreadCrumb = function( eventData ) {
    var output = {};
    const searchCriteria = AwStateService.instance.current.data && AwStateService.instance.current.data.params ? AwStateService.instance.current.data.params : '';
    const label = AwStateService.instance.current.data && AwStateService.instance.current.data.label ? AwStateService.instance.current.data.label : '';
    output.breadcrumbConfig = appCtxService.getCtx( 'breadCrumbConfig' );
    output.breadCrumbProvider = breadCrumbService.refreshBreadcrumbProvider( output.breadcrumbConfig,
        appCtxService.getCtx( 'mselected' ),
        eventData.searchFilterCategories, eventData.searchFilterMap,
        searchCriteria, label, true );

    return output;
};

export let resetBreadCrumb = function() {
    var output = {};
    output.breadcrumbConfig = appCtxService.getCtx( 'breadCrumbConfig' );
    output.breadCrumbProvider = breadCrumbService.resetBreadcrumbProvider( output.breadcrumbConfig );
    return output;
};

export let updateDocumentTitles = function() {
    if( appCtxService.ctx.location ) {
        document.title = appCtxService.ctx.location.titles.browserTitle +
             ( appCtxService.ctx.location.titles.browserSubTitle ? ' - ' + appCtxService.ctx.location.titles.browserSubTitle : '' );
    }
};

export let constructTabs = function( subPages ) {
    var subLocationTabs = [];
    var promises = [];

    var constructTabFromState = function( name, pageId, priority, isSelected, stateName, selectWhen, isCloseable ) {
        return {
            classValue: 'sw-tab-title',
            name: name,
            tabKey: name,
            displayTab: true,
            pageId: pageId,
            priority: priority,
            selectedTab: isSelected,
            state: stateName,
            selectWhen: selectWhen,
            visible: true,
            closeable: isCloseable,
            closeCommandIcon: 'cmdCloseTab',
            closeCallback: 'closeTab',
            closeCommandTitle: 'Close Tab'
        };
    };

    subPages.sort( function( obj1, obj2 ) {
        return obj1.data.priority - obj2.data.priority;
    } );

    _.forEach( subPages, function( page, index ) {
        var label = page.data.label;
        var isCloseable = page.data.closeable;
        var isSelectedTab = page.url === AwStateService.instance.current.url;
        var priority = page.data.priority ? page.data.priority : 0;
        var stateName = page.name;
        if( label ) {
            var selectWhen = 'data.subLocationTabCond.currentTab === \'' + stateName + '\'';
            if( _.isString( label ) ) {
                subLocationTabs.push( constructTabFromState( label, index, priority, isSelectedTab, stateName, selectWhen, isCloseable ) );
                promises.push( AwPromiseService.instance.when() );
            } else {
                promises.push( localeService.getLocalizedText( label.source, label.key ).then( function( result ) {
                    subLocationTabs.push( constructTabFromState( result, index, priority, isSelectedTab, stateName, selectWhen, isCloseable ) );
                } ) );
            }
        }
    } );

    return AwPromiseService.instance.all( promises ).then( function() {
        return subLocationTabs;
    } );
};

export let constructTabsAndGetActiveTab = ( subPages ) => {
    return constructTabs( subPages ).then( function( subLocationTabs ) {
        return {
            subLocationTabs: subLocationTabs,
            activeTab: getActiveSublocationTab( subLocationTabs )
        };
    } );
};

export let switchSubLocation = function( pageId, tabTitle, tabsModel, data ) {
    var title = tabTitle || pageId;
    if( title && data && title === data.primaryActiveTab ) {
        return;
    }
    const { dispatch } = data;
    var tabToSelect = tabsModel.find( function( tab ) {
        return tab.name === title;
    } );

    if( tabToSelect ) {
        // When the tab widget is forced to update after the state has already changed it will still trigger callback
        if( tabToSelect.state !== AwStateService.instance.current.name ) {
            dispatch && dispatch( { path: 'data.primaryActiveTab', value: title } );
            if( tabToSelect.params ) {
                AwStateService.instance.go( tabToSelect.state, tabToSelect.params );
            } else {
                AwStateService.instance.go( tabToSelect.state, AwStateService.instance.params );
            }
        }
    } else {
        logger.error( 'Missing tab was selected: ' + tabTitle );
    }
};
export let closeTab = function( subLocationTabs, removedTab ) {
    if( removedTab !== undefined ) {
        const index = subLocationTabs.indexOf( removedTab );
        if( index >= 0 ) {
            subLocationTabs.splice( index, 1 );
        }
    }
};

const awHeaderService = {
    getTitles,
    updateBreadCrumb,
    resetBreadCrumb,
    updateDocumentTitles,
    constructTabs,
    constructTabsAndGetActiveTab,
    switchSubLocation,
    getBreadcrumbConfig,
    closeTab
};
export default awHeaderService;
