// @swf/core/src/plstats/src/js/splmStatsDOMProcessor.js
// Copyright (c) 2020 Siemens
/* eslint-env es6 */

/**
 * Dom Processor - Provides performance metrics in relation to the DOM
 * NOTE: this doesn't have full functionality as the afx version!
 *
 * @module js/splmStatsDOMProcessor
 */
import splmStatsUtils from 'js/splmStatsUtils';

/**
 * Instances of this class represent a profiler for DOM Information
 *
 * @class SPLMStatsDomProcessor
 */
function SPLMStatsDomProcessor() {
    let self = this;

    let _processingTime = 0;

    let _startProcessorTime = 0;

    let _endProcessorTime = 0;
    /**
     * calculating Processing Time
     */
    function calculateProcessingTime() {
        _startProcessorTime = window.performance.now();
        _endProcessorTime = window.performance.now();
        _processingTime += _endProcessorTime - _startProcessorTime;
    }

    self.start = function() {
        calculateProcessingTime();
    };
    self.stop = function() {
        calculateProcessingTime();
    };

    /**
     * @returns {Object} DOM Object contains element count on page, DOM Tree Depth, and # of costly widgets on page
     * Definition of costly: N >= 6 depth OR N >= 50 watchers on the DOM structure ( element and children )
     */
    self.getMetrics = function() {
        return {
            DOM: {
                elemCount: splmStatsUtils.getDomElementsCount(),
                DOMTreeDepth: splmStatsUtils.getDomTreeDepth( document, 0 )
                /*
                DOMCostlyWidgets: splmStatsUtils.getCostlyWidgets()
                */
            }
        };
    };

    /**
     * @returns {Object} Total processing time for DOM Metrics
     */
    self.getProcessingTime = function() {
        const _time = _processingTime;
        _processingTime = 0;
        return { DOMProcessorOverhead: _time };
    };

    return self;
}

export default SPLMStatsDomProcessor;
