// @swf/core/src/kernel/src/js/appWrapper.js
// Copyright (c) 2022 Siemens

/**
 * This module provides APIs common to initialize and interact with AngularJS-based application modules.
 *
 * @module app
 * @deprecated @swf/core@6.0.0
 */

/**
 * Adds an AngularJS 'controller' definition to this application.
 * 
 * @deprecated @swf/core@6.0.0
 */
export let controller = function() {
    // do nothing
};

/**
 * Adds an AngularJS 'directive' definition to this application.
 * 
 * @deprecated @swf/core@6.0.0
 */
export let directive = function() {
    // do nothing
};

/**
 * Adds an AngularJS 'component' definition to this application.
 * 
 * @deprecated @swf/core@6.0.0
 */
export let component = function() {
    // do nothing
};

/**
 * Adds an AngularJS 'filter' definition to this application.
 * 
 * @deprecated @swf/core@6.0.0
 */
export let filter = function() {
    // do nothing
};

/**
 * Adds an AngularJS injectable 'service' definition to this application.
 * 
 * @deprecated @swf/core@6.0.0
 */
export let factory = function() {
    // do nothing
};

/**
 * Adds an AngularJS injectable 'service' for this application.
 * 
 * @deprecated @swf/core@6.0.0
 */
export let service = function() {
    // do nothing
};

/**
 * Adds an AngularJS injectable 'constant' for this application.
 *
 * @deprecated @swf/core@6.0.0
 */
export let constant = function() {
    // do nothing
};

/**
 * Get the base url path.
 *
 * @returns {String} The relative path from the root of the deployed war file to all modules and resources used by
 *          this application module (e.g. 'assets').
 * @deprecated @swf/core@6.0.0
 */
export let getBaseUrlPath = function() {
    return 'assets';
};

export default {
    service,
    directive,
    controller,
    component,
    filter,
    factory,
    getBaseUrlPath,
    constant
};
