// @swf/core/src/uiwidget/test/mocks/mockIndicators.js
// Copyright (c) 2020 Siemens
/* eslint-env es6 */

/**
 * @module test/mockIndicators
 */

export default {
    indicator_1: {
        iconName: 'icon1',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_1' ]
        },
        modelTypes: [ 'myFakeObjectType' ],
        conditions: {
            fakePropertyName_1: {
                $eq: 'test'
            }
        }
    },
    indicator_2: {
        iconName: 'icon2',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_2' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_2: {
                $eq: '2'
            }
        }
    },
    validSimpleRegexpIndicator: {
        iconName: 'icon3',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_3' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_3: {
                $regexp: '^[a-z0-9,]+$'
            }
        }
    },
    validComplexRegexpIndicator: {
        iconName: 'icon3',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_3' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_3: {
                $regexp: {
                    $pattern: '^[a-z]+$',
                    $options: 'i'
                }
            }
        }
    },
    invalidComplexRegexpIndicator: {
        iconName: 'icon3',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_3' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_3: {
                $regexp: {
                    $patterns: '^[a-z]+$'
                }
            }
        }
    },
    invalidComplexRegexpIndicator2: {
        iconName: 'icon3',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_3' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_3: {
                $regexp: {
                    $pattern: '^[a-z]+$',
                    $options: '!!!'
                }
            }
        }
    },
    validLtIndicator: {
        iconName: 'icon3',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_3' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_3: {
                $lt: '3'
            }
        }
    },
    inValidLtIndicator: {
        iconName: 'icon3',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_3' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_3: {
                $lt: 'test'
            }
        }
    },
    validLteIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $lte: '3'
            }
        }
    },
    inValidNumberLteIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $lte: '4'
            }
        }
    },
    inValidStringLteIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $lte: 'test'
            }
        }
    },
    validGtIndicator: {
        iconName: 'icon5',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_5' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_5: {
                $gt: '2'
            }
            // check if number else fail
        }
    },
    inValidGtIndicator: {
        iconName: 'icon6',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_6' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_6: {
                $gt: 'test'
            }
        }
    },
    validGteIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $gte: '3'
            }
        }
    },
    inValidNumberGteIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $gte: '2'
            }
        }
    },
    inValidStringGteIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $gte: 'test'
            }
        }
    },
    inValidConditionFormat: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName: 'test'
        }
    },
    inValidConditionOperator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName: {
                not: 'testing'
            }
        }
    },
    indicator_7: {
        iconName: 'icon7',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_7' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_7: {
                $ne: ''
            }
            // property dbValue
        }
    },

    // EQ MOCK DATA
    validEqIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 'test'
            }
        }
    },
    validEqStringIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 'test'
            }
        }
    },
    validEqIntIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 3
            }
        }
    },
    validEqStringToIntCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: '3'
            }
        }
    },
    validEqIntToStringCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 3
            }
        }
    },
    validEqBooleanCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: true
            }
        }
    },
    invalidEqBooleanCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: false
            }
        }
    },
    invalidEqIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 'nope'
            }
        }
    },
    invalidEqStringIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 'nope'
            }
        }
    },
    invalidEqIntIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 30
            }
        }
    },
    invalidEqStringToIntCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: '30'
            }
        }
    },
    invalidEqIntToStringCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $eq: 30
            }
        }
    },

    // NE MOCK DATA
    validNeIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 'notTest'
            }
        }
    },
    validNeStringIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 'notTest'
            }
        }
    },
    validNeIntIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 4
            }
        }
    },
    validNeBooleanCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: false
            }
        }
    },
    invalidNeBooleanCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: true
            }
        }
    },
    validNeStringToIntCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: '4'
            }
        }
    },
    validNeIntToStringCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 4
            }
        }
    },
    invalidNeIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 'test'
            }
        }
    },
    invalidNeStringIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 'test'
            }
        }
    },
    invalidNeIntIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 3
            }
        }
    },
    invalidNeStringToIntCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: '3'
            }
        }
    },
    invalidNeIntToStringCompareIndicator: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [ 'fakePropertyName_4' ]
        },
        modelTypes: [ 'myFakeObjectType', 'myFakeObjectTypeParent' ],
        conditions: {
            fakePropertyName_4: {
                $ne: 3
            }
        }
    },
    // Multiple Indicator Mockdata
    multiple_Indicator_pending: {
        iconName: 'icon1',
        tooltip: {
            showPropDisplayName: false,
            propNames: [
                'object_string'
            ]
        },
        prop: {
            names: [
                'multipleIndicatorProperty1'
            ],
            type: {
                names: [
                    'myFakeObjectType'
                ],
                prop: {
                    names: [
                        'object_string'
                    ],
                    conditions: {
                        object_string: {
                            $eq: 'Pending'
                        }
                    }
                }
            }
        },
        modelTypes: [
            'myFakeObjectTypeParent'
        ]
    },
    multiple_Indicator_default: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [
                'object_string'
            ]
        },
        prop: {
            names: [
                'multipleIndicatorProperty1'
            ],
            conditions: {}
        },
        modelTypes: [
            'myFakeObjectTypeParent'
        ]
    },
    // Multiple Indicator Mockdata 2
    multiple_Indicator_pending2: {
        iconName: 'icon1',
        tooltip: {
            showPropDisplayName: false,
            propNames: [
                'object_string'
            ]
        },
        prop: {
            names: [
                'multipleIndicatorProperty2'
            ],
            type: {
                names: [
                    'myFakeObjectType'
                ],
                prop: {
                    names: [
                        'object_string'
                    ],
                    conditions: {
                        object_string: {
                            $eq: 'Pending2'
                        }
                    }
                }
            }
        },
        modelTypes: [
            'myFakeObjectTypeParent'
        ]
    },
    multiple_Indicator_default2: {
        iconName: 'icon4',
        tooltip: {
            showPropDisplayName: false,
            propNames: [
                'object_string'
            ]
        },
        prop: {
            names: [
                'multipleIndicatorProperty2'
            ],
            conditions: {}
        },
        modelTypes: [
            'myFakeObjectTypeParent'
        ]
    },
    // Displaying tooltip using tooltip propNames mock data
    multiple_Indicator_approved: {
        iconName: 'icon8',
        tooltip: {
            showPropDisplayName: false,
            propNames: [
                'object_name',
                'date_released'
            ]
        },
        prop: {
            names: [
                'object_name'
            ],
            type: {
                names: [
                    'myFakeObjectType'
                ],
                prop: {
                    names: [
                        'object_name'
                    ],
                    conditions: {
                        object_name: {
                            $eq: 'Approved'
                        }
                    }
                }
            }
        },
        modelTypes: [
            'myFakeObjectTypeParent'
        ]
    },
    mock_release_status_list_tcm_released: {
        iconName: 'icon9',
        tooltip: {
            showPropDisplayName: false,
            propNames: [
                'object_name',
                'date_released'
            ]
        },
        prop: {
            names: [
                'release_status_list'
            ],
            type: {
                names: [
                    'ReleaseStatus'
                ],
                prop: {
                    names: [
                        'object_name'
                    ],
                    conditions: {
                        object_name: {
                            $eq: 'TCM Released'
                        }
                    }
                }
            }
        },
        modelTypes: [
            'myFakeObjectTypeParent'
        ]
    }
};
