// @swf/core/src/uiwidget/test/mocks/mockDecorators.js
// Copyright (c) 2020 Siemens
/* eslint-env es6 */

/**
 * @module test/mockDecorators
 */

export default {
    decorator_1: {
        cellClassName: 'aw-border-chartColor1',
        gridClassName: 'aw-charts-chartColor1',
        modelTypes: [ 'myFakeObjectType' ],
        conditions: {
            fakePropertyName_1: {
                expression: {
                    $source: 'localContext.vmo.props.fakePropertyName_1.dbValue',
                    $query: {
                        $eq: 'fakeDBValue1'
                    }
                }
            }
        }
    },
    decorator_2: {
        cellClassName: 'aw-border-chartColor2',
        gridClassName: 'aw-charts-chartColor2',
        conditions: {
            fakePropertyName_1: {
                expression: {
                    $source: 'localContext.vmo.props.fakePropertyName_1.dbValue',
                    $query: {
                        $eq: 'fakeDBValue2'
                    }
                }
            }
        }
    }
};
