// @swf/tooling/view/compiler.js
// Copyright (c) 2020 Siemens
/* eslint-env node */
const createCompiler = require( './compilerFactory' );

// compiler definition
const _compiler = createCompiler( [
    // custom attributes
    require( './compileAwRepeat' ),
    require( './compileExistWhen' ),
    require( './compileVisibleWhen' ),
    require( './compileEnableWhen' ),
    require( './compileExtendedTooltip' ),
    require( './compileAwClick' ),
    require( './compileAwRightClick' ),
    require( './compileAwClass' ),
    require( './compileAwEnterKey' ),
    require( './compileAwAutofocus' ),
    require( './compileAwEnter' ),
    require( './compileAwLongPress' ),
    require( './compileShowWhen' ),
    require( './compileAwHighlightPropertyHtml' ),
    require( './compileAwDependOn' ),

    // custom elements
    require( './compileAwDefaultGroupCell' ),
    require( './compileAwPanel' ),
    require( './compileAwCommandPanel' ),
    require( './compileAwList' ),
    require( './compileAwFileUpload' ),
    require( './compileAwSlider' ),
    require( './compileAwTabSet' ),
    require( './compileAwCheckList' ),
    require( './compileAwGraph' ),
    require( './compileAwChart' ),
    require( './compileAwChip' ),
    require( './compileAwChipList' ),
    require( './compileAwTree' ),
    require( './compileAwInclude' ),
    require( './compileAwTileCanvas' ),

    // popup related - replaced by build scripts 'build\js\migratePopup.js'
    // require( './compilePopup' ),

    // general compile
    require( './compileElement' ),
    require( './compileText' )
    // require( './compileComments' )
] );

const compile = ( parsedView, popupCompileContext = null ) => {
    return _compiler.compile( parsedView, {
        compileFn: _compiler.compile,
        popupCompileContext,
        locals: {
            props: true,
            data: true,
            fields: true,
            dataProviders: true,
            ctx: true,
            i18n: true,
            actions: true,
            subPanelContext: true,
            viewPath: true,
            conditions: true,
            formProp: true,
            dispatch: true,
            selectionModels: true,
            editHandlers: true,
            chartProviders: true,
            ports: true
        }
    } ).join( '\n' );
};

module.exports = compile;
