// @swf/core/src/command/test/AwCellCommandListTestData.js
export default {
    cellCommandList: [ {
        id: 'cellCmd',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'cellCmd',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    topRightCellCommandList: [ {
        id: 'topRightCmd',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'topRightCmd',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE',
        cellDisplay : {
            position: 'TOP_RIGHT'
        }
    } ],
    manualStyleCellCommandList: [ {
        id: 'manualStyleCmd',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'manualStyleCmd',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE',
        cellDisplay : {
            position: 'MANUAL_STYLE'
        }
    } ],
    middleRightCellCommandList: [ {
        id: 'middleRightCmd',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'middleRightCmd',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE',
        cellDisplay : {
            position: 'MIDDLE_RIGHT'
        }
    } ],
    bottomRightCellCommandList: [ {
        id: 'bottomRightCmd',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'bottomRightCmd',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE',
        cellDisplay : {
            position: 'BOTTOM_RIGHT'
        }
    } ]
};
