// @swf/core/src/command/test/AwCommandListTestData.js
export default {
    baseCommandEnableState: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    baseCommandComplexCondition: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: {
                    $source: '(ctx.selected && ctx.mselected.length === 1)',
                    $query: {
                        $eq: true
                    },
                    __useSourceParsing: true
                }
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    baseCommandSelectState: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            selectWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    baseCommandIdSelectState: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    baseCommandVisibleState: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            visibleWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    baseCommandMultipleHandlers: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            priority: 5,
            visibleWhen: {
                id: 'false',
                expression: 'false'
            }
        }, {
            id: 'myCommandHandler2',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    baseCommandNoHandlers: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    commandWithDynamicTitle: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command {0}',
            parameters: [ '{{ctx.myThingFromCtx}}' ]
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    commandWithDynamicTitleMultipleParameters: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command {0} {1}',
            parameters: [ '{{ctx.myThingFromCtx}}', '{{ctx.myThingFromCtx2}}' ]
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    commandWithDynamicSelectedTitle: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            },
            selectWhen: {
                id: 'condition',
                expression: 'ctx.sFlag'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: {
            value: 'My Command {0}',
            parameters: [ '{{ctx.myThingFromCtx}}' ]
        },
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    commandWithDynamicTitleCommandContext: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command {0}',
            parameters: [ '{{commandContext.myThingFromCtx}}' ]
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    commandWithDynamicSelectedTitleCommandContext: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            enableWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            },
            visibleWhen: {
                id: 'true',
                expression: 'true'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: {
            value: 'My Command {0}',
            parameters: [ '{{commandContext.myThingFromCtx}}' ]
        },
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    groupCommandWithDynamicTitle: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Command {0}',
            parameters: [ '{{ctx.myThingFromCtx}}' ]
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    groupCommandWithDynamicSelectedTitle: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Group Command',
            parameters: []
        },
        selectedTitle: {
            value: 'My Command {0}',
            parameters: [ '{{ctx.myThingFromCtx}}' ]
        },
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    groupCommandWithDynamicTitleCommandContext: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Command {0}',
            parameters: [ '{{commandContext.myThingFromCtx}}' ]
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'true'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'true'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'true'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    groupCommandWithDynamicSelectedTitleCommandContext: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Group Command',
            parameters: []
        },
        selectedTitle: {
            value: 'My Command {0}',
            parameters: [ '{{commandContext.myThingFromCtx}}' ]
        },
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'true',
                    expression: 'true'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'true'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'true'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    commandsCheckingSeparateFlags: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            visibleWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    }, {
        id: 'myCommand2',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            visibleWhen: {
                id: 'condition',
                expression: 'ctx.flag2'
            }
        } ],
        title: {
            value: 'My Command 2',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ],
    groupCommandSingleChild: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Group Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    groupCommandSingleChildNoSelected: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Group Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        showGroupSelected: false,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    groupCommandMultipleChild: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Group Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'GROUP',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        }, {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag2'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag2'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag2'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    toggleBaseCommand: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            selectWhen: {
                id: 'condition',
                expression: 'ctx.sFlag'
            },
            visibleWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            }
        } ],
        title: {
            value: 'myTitle',
            parameters: []
        },
        // object like title
        // if not selected title.value = My toggle base
        // if selected title.value include myselectedtitle
        selectedTitle: {
            value: 'mySelectedTitle',
            parameters: []
        },
        icon: 'myIcon',
        selectedIcon: 'mySelectedIcon',
        description: null,
        type: 'BASE'
    } ],
    toggleRibbonCommand: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            selectWhen: {
                id: 'condition',
                expression: 'ctx.SFlag'
            },
            visibleWhen: {
                id: 'condition',
                expression: 'ctx.flag'
            }
        } ],
        title: {
            value: 'myTitle',
            parameters: []
        },
        selectedTitle: {
            value: 'mySelectedTitle',
            parameters: []
        },
        icon: 'myIcon',
        selectedIcon: 'mySelectedIcon',
        description: null,
        type: 'RIBBON'
    } ],
    shuttleCommandSingleChild: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Shuttle Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'SHUTTLE',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    shuttleCommandSingleChildNoSelected: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Shuttle Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        showGroupSelected: false,
        type: 'SHUTTLE',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    shuttleCommandMultipleChild: [ {
        id: 'myCommand',
        handlers: [],
        title: {
            value: 'My Shuttle Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'SHUTTLE',
        children: [ {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        }, {
            id: 'myCommand',
            handlers: [ {
                id: 'myCommandHandler',
                activeWhen: {
                    id: 'true',
                    expression: 'true'
                },
                priority: 4,
                visibleWhen: {
                    id: 'condition',
                    expression: 'ctx.flag2'
                },
                enableWhen: {
                    id: 'condition',
                    expression: 'ctx.eFlag2'
                },
                selectWhen: {
                    id: 'condition',
                    expression: 'ctx.sFlag2'
                }
            } ],
            title: {
                value: 'My Command',
                parameters: []
            },
            selectedTitle: null,
            icon: 'myIcon',
            selectedIcon: null,
            description: null,
            type: 'BASE'
        } ]
    } ],
    baseCommandWithContext: [ {
        id: 'myCommand',
        handlers: [ {
            id: 'myCommandHandler',
            activeWhen: {
                id: 'true',
                expression: 'true'
            },
            priority: 4,
            visibleWhen: {
                id: 'condition',
                expression: 'commandContext.flag'
            }
        } ],
        title: {
            value: 'My Command',
            parameters: []
        },
        selectedTitle: null,
        icon: 'myIcon',
        selectedIcon: null,
        description: null,
        type: 'BASE'
    } ]
};
