// @swf/core/src/command/test/commandConfigurationServiceTestData.js
export default {
    commandWithDynamicTitle: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchor',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchor',
                priority: 2
            }
        },
        commands: {
            CmdP1: {
                title: {
                    text: '{0}',
                    params: [
                        '{{ctx.previousLocationDisplayName}}'
                    ]
                }
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            }
        }
    },
    commandsWithPriority123: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchor',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchor',
                priority: 2
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            }
        }
    },
    commandsWithMissingPriority23: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor'
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchor',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchor',
                priority: 2
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            }
        }
    },
    commandsWithP1AndRAn1: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchor',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchor',
                relativeTo: 'CmdP1',
                priority: -1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            }
        }
    },
    commandsWithP1AndRA1: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchor',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchor',
                relativeTo: 'CmdP1',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            }
        }
    },
    commandWithBadRelativeTo: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchor',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchor',
                relativeTo: 'MyMissingCommand',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            }
        }
    },
    shuttleCommand: {
        commandHandlers: {},
        actions: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isShuttle: true,
                isGroup: true
            }
        }
    },
    ribbonCommand: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isRibbon: true,
                isGroup: true
            }
        }
    },
    groupCommand: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isGroup: true
            }
        },
        actions: {}
    },
    toggleCommand: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isToggle: true
            }
        }
    },
    baseCommand: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            }
        }
    },
    commandSimpleTitle: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'My Title'
            }
        }
    },
    commandI18nTitle: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: '{{i18n.title}}'
            }
        },
        i18n: {
            title: [
                'MyMessage'
            ]
        }
    },
    commandSimpleSelectedTitle: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'My Title',
                selected: {
                    title: 'My Selected Title'
                }
            }
        }
    },
    commandI18nSelectedTitle: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'My Title',
                selected: {
                    title: '{{i18n.title}}'
                }
            }
        },
        i18n: {
            title: [
                'MyMessage'
            ]
        }
    },
    commandI18nMissingTitle: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: '{{i18n.title}}'
            }
        },
        i18n: {}
    },
    commandObjectTitleSimple: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: {
                    text: 'My Title',
                    params: [ 'ctx.myParameter' ]
                }
            }
        }
    },
    commandObjectTitleI18n: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: {
                    text: '{{i18n.title}}',
                    params: [ 'ctx.myParameter' ]
                }
            }
        },
        i18n: {
            title: [
                'MyMessage'
            ]
        }
    },
    ribbonCommandNoActionHandler: {
        actions: {
            openCommandGroup: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'AwPopupCommandBar'
                    }
                }
            }
        },
        commandHandlers: {
            ribbonHandler: {
                id: 'CmdP1',
                activeWhen: {
                    condition: 'conditions.true'
                },
                visibleWhen: {
                    condition: 'conditions.true'
                }
            }
        },
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isRibbon: true,
                isGroup: true
            }
        },
        conditions: {
            true: {
                expression: 'true'
            },
            false: {
                expression: 'false'
            }
        }
    },
    ribbonCommandWithActionHandler: {
        actions: {
            openCommandGroup: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'AwPopupCommandBar'
                    }
                }
            },
            myAction: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'MyView'
                    }
                }
            }
        },
        commandHandlers: {
            ribbonHandler: {
                id: 'CmdP1',
                action: 'myAction',
                activeWhen: {
                    condition: 'conditions.true'
                },
                visibleWhen: {
                    condition: 'conditions.true'
                }
            }
        },
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isRibbon: true,
                isGroup: true
            }
        },
        conditions: {
            true: {
                expression: 'true'
            },
            false: {
                expression: 'false'
            }
        }
    },
    groupCommandWithChild: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            },
            c1: {
                id: 'CmdC1',
                uiAnchor: 'CmdP1',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isGroup: true
            },
            CmdC1: {
                title: 'My Title'
            }
        },
        actions: {
            openCommandGroup: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'AwPopupCommandBar'
                    }
                }
            }
        }
    },
    handlerWithServerVisibility: {
        actions: {
            openCommandGroup: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'AwPopupCommandBar'
                    }
                }
            },
            myAction: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'MyView'
                    }
                }
            }
        },
        commandHandlers: {
            ribbonHandler: {
                id: 'CmdP1',
                action: 'myAction',
                activeWhen: {
                    condition: 'conditions.true'
                },
                visibleWhen: {
                    condition: 'conditions.serverVisibilityCondition'
                }
            }
        },
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isRibbon: true,
                isGroup: true
            }
        },
        conditions: {
            true: {
                expression: 'true'
            },
            false: {
                expression: 'false'
            },
            serverVisibilityCondition: {
                expression: 'ctx.visibleServerCommands.myCommandId'
            }
        }
    },
    commandWithIcon: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                iconId: 'myIcon',
                selected: {
                    iconId: 'mySelectedIcon'
                }
            }
        }
    },
    commandWithMultipleHandlers: {
        actions: {
            myAction: {
                actionType: 'popup',
                inputData: {
                    options: {
                        view: 'MyView'
                    }
                }
            }
        },
        commandHandlers: {
            handler1: {
                id: 'CmdP1',
                action: 'myAction',
                activeWhen: {
                    condition: 'conditions.true'
                },
                visibleWhen: {
                    condition: 'conditions.true'
                }
            },
            handler2: {
                id: 'CmdP1',
                action: 'myAction',
                activeWhen: {
                    condition: 'conditions.doubleTrue'
                },
                visibleWhen: {
                    condition: 'conditions.true'
                }
            }
        },
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            }
        },
        conditions: {
            doubleTrue: {
                expression: 'true && true'
            },
            true: {
                expression: 'true'
            },
            false: {
                expression: 'false'
            }
        }
    },
    groupCommandNoPlacement: {
        commandHandlers: {},
        commandPlacements: {},
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isGroup: true
            }
        },
        actions: {}
    },
    groupCommandNoShowGroupSelected: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isGroup: true
            }
        },
        actions: {}
    },
    groupCommandShowGroupSelectedFalse: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchor',
                priority: 1,
                showGroupSelected: false
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1',
                isGroup: true
            }
        },
        actions: {}
    },
    baseCommandNoPlacement: {
        commandHandlers: {},
        commandPlacements: {},
        commands: {
            CmdP1: {
                title: 'CmdP1'
            }
        }
    },
    cellCommandBar: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'topRightCmd',
                uiAnchor: 'anchor',
                priority: 1,
                cellDisplay: {
                    hover: true,
                    position: 'TOP_RIGHT'
                }
            },
            p2: {
                id: 'manualStyleCmd',
                uiAnchor: 'anchor',
                priority: 2,
                cellDisplay: {
                    hover: true,
                    position: 'MANUAL_STYLE'
                }
            },
            p3: {
                id: 'middleRightCmd',
                uiAnchor: 'anchor',
                priority: 3,
                cellDisplay: {
                    hover: true,
                    position: 'MIDDLE_RIGHT'
                }
            },
            p4: {
                id: 'bottomRightCmd',
                uiAnchor: 'anchor',
                priority: 4,
                cellDisplay: {
                    hover: true,
                    position: 'BOTTOM_RIGHT'
                }
            },
            p5: {
                id: 'cellCmd',
                uiAnchor: 'anchor',
                priority: 5
            }
        },
        commands: {
            cellCmd: {
                title: 'cellCmd'
            },
            topRightCmd: {
                title: 'topRightCmd'
            },
            manualStyleCmd: {
                title: 'manualStyleCmd'
            },
            middleRightCmd: {
                title: 'middleRightCmd'
            },
            bottomRightCmd: {
                title: 'bottomRightCmd'
            }
        },
        actions: {}
    },
    commandsWithMultipleAnchors: {
        commandHandlers: {},
        commandPlacements: {
            p1: {
                id: 'CmdP1',
                uiAnchor: 'anchorP',
                priority: 1
            },
            p3: {
                id: 'CmdP3',
                uiAnchor: 'anchorP',
                priority: 3
            },
            p2: {
                id: 'CmdP2',
                uiAnchor: 'anchorP',
                priority: 2
            },
            x1: {
                id: 'CmdX1',
                uiAnchor: 'anchorX',
                priority: 1
            },
            x3: {
                id: 'CmdX3',
                uiAnchor: 'anchorX',
                priority: 3
            },
            x2: {
                id: 'CmdX2',
                uiAnchor: 'anchorX',
                priority: 2
            }
        },
        commands: {
            CmdP1: {
                title: 'CmdP1'
            },
            CmdP2: {
                title: 'CmdP2'
            },
            CmdP3: {
                title: 'CmdP3'
            },
            CmdX1: {
                title: 'CmdX1'
            },
            CmdX2: {
                title: 'CmdX2'
            },
            CmdX3: {
                title: 'CmdX3'
            }
        }
    }
};
