// @swf/core/src/declreact/src/js/hocCollection.js
// Copyright (c) 2021 Siemens

/**
 * @module js/hocCollection
 */
export { default as awInitHOC } from 'js/awInitHOC';
export { default as ExtendedTooltip } from 'js/extendedTooltipHOC';
export { default as ExtTooltipData } from 'js/extendedTooltipHOC';
export { default as VisibleWhen } from 'js/awVisibleWhenHOC';
export { default as EnableWhen } from 'js/awEnableWhenHOC';
export { default as ExistWhen } from 'js/awExistWhenHOC';
export { default as AwClick } from 'js/awClickHOC';
export { default as AwClickOptions } from 'js/awClickHOC';
export { default as AwRightClick } from 'js/awRightClickHOC';
export { default as AwClass } from 'js/awClassHOC';
export { default as AwEnterKey } from 'js/awEnterKeyHOC';
export { default as AwAutofocus } from 'js/awAutofocusHOC';
export { default as AwEnter } from 'js/awEnterHOC';
export { default as AwLongPress } from 'js/awLongPressHOC';
export { default as AwHighlightPropertyHtml } from 'js/awHighlightPropertyHtmlHOC';
export { default as DisplayVal } from 'js/awHighlightPropertyHtmlHOC';
export { default as AwParseHtml } from 'js/awHighlightPropertyHtmlHOC';
export { default as ShowWhen } from 'js/awShowWhenHOC';
export { default as DependOn } from 'js/awDependOnHOC';
export { default as OnChange } from 'js/awDependOnHOC';
